content/gocourse-2019-05-melbourne.slide

```diff
@@ -1003,6 +1003,43 @@ Use `gofmt` or `goimport` as part of your editor/IDE setup
 - ...
 
 
+* JSON - (De-)Serialisation
+
+- `Encode` converts Go struct, pointer or primitive data type into JSON encoded stream
+- `Marshal` converts Go struct, pointer or primitive data type into JSON byte array
+- `Decode` and `Unmarshall` are the inverse functions to `Encode` and `Marshal`
+
+	func Marshal(v interface{}) ([]byte, error) 
+
+	func Unmarshal(data []byte, v interface{}) error 
+
+
+* JSON - Tags
+
+- Attach `json` tag to specify JSON property name
+- Common error is making Go fields private (lower case) which results in the fields not being (un)marshalled
+
+
+	type Car struct {
+		Manufacturer string `json:"manufacturer"`
+		Model        string `json:"model"`
+		Year         int64  `json:"year,omitempty"`
+	}
+
+
+* JSON - Zero Values
+
+- All Go struct fields will be set to their type's zero value if the corresponding JSON property is missing.
+- Pointer types have a zero value of `nil`
+- To be able to differentiate between a field which has been set explicitly to it's zero value or omitted, you must use pointer types
+
+
+* Examples
+
+- [[https://goplay.space/#_LVQLCOKYNT][JSON Marshal / Unmarshal]]
```

Comment:
https://goplay.space/#lYdq4Jg19EP

I've added another `Person` Marshal that has `FirstName` but not `LastName` to show what happens on fields without `omitempty` as a contrast to the prior `Person` Marshal.

Reply:
I like it, I've added this in now and updated the associated in `samples`